File: packages/node/src/bridge/protocol.ts
import crypto from 'node:crypto';
import type { BridgeMethod, BridgeProgressParams, BridgeRequest, BridgeResponse } from '../types/bridge.js';
import { BridgeErrorCode, createError, type LionScraperError } from '../types/errors.js';
import { t, type MessageId, type SupportedLang } from '../i18n/lang.js';
import { logger } from '../utils/logger.js';

/** PendingRequest.rejectAll reason → `t()` message id for BRIDGE_DISCONNECTED. */
export type DisconnectRejectId = Extract<MessageId, `disconnect.${string}`>;

export function createBridgeRequest(
  method: BridgeMethod,
  params?: Record<string, unknown>,
): BridgeRequest {
  return {
    jsonrpc: '2.0',
    id: crypto.randomUUID(),
    method,
    ...(params && { params }),
  };
}

export function createBridgeResponse(id: string, result: unknown): BridgeResponse {
  return { jsonrpc: '2.0', id, result };
}

export function createBridgeErrorResponse(
  id: string,
  code: number,
  message: string,
  data?: Record<string, unknown>,
): BridgeResponse {
  return {
    jsonrpc: '2.0',
    id,
    error: { code, message, ...(data && { data }) },
  };
}

export type BridgeProgressHandler = (params: BridgeProgressParams) => void;

export interface PendingRequest {
  id: string;
  method: BridgeMethod;
  resolve: (value: unknown) => void;
  reject: (reason: LionScraperError) => void;
  timer: ReturnType<typeof setTimeout>;
  createdAt: number;
  lang: SupportedLang;
  onProgress?: BridgeProgressHandler;
}

const DEFAULT_TIMEOUT_MS = 60_000;

export class PendingRequestManager {
  private pending = new Map<string, PendingRequest>();

  add(
    id: string,
    method: BridgeMethod,
    timeoutMs: number = DEFAULT_TIMEOUT_MS,
    lang: SupportedLang = 'en-US',
    onProgress?: BridgeProgressHandler,
  ): Promise<unknown> {
    return new Promise<unknown>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        logger.warn(`Bridge request timed out: ${id} (${method})`);
        reject(
          createError(
            BridgeErrorCode.BRIDGE_TIMEOUT,
            t(lang, 'bridge_timeout', { ms: timeoutMs }),
          ),
        );
      }, timeoutMs);

      this.pending.set(id, {
        id,
        method,
        resolve,
        reject,
        timer,
        createdAt: Date.now(),
        lang,
        onProgress,
      });
    });
  }

  /**
   * Forward Extension → Server bridgeProgress notification to the matching pending tool call.
   * @returns true if a pending request consumed this payload
   */
  dispatchProgress(requestId: string, params: BridgeProgressParams): boolean {
    const pending = this.pending.get(requestId);
    if (!pending) return false;
    try {
      pending.onProgress?.(params);
    } catch {
      /* ignore handler errors */
    }
    return true;
  }

  resolve(id: string, result: unknown): boolean {
    const pending = this.pending.get(id);
    if (!pending) return false;

    clearTimeout(pending.timer);
    this.pending.delete(id);
    pending.resolve(result);
    return true;
  }

  reject(id: string, error: LionScraperError): boolean {
    const pending = this.pending.get(id);
    if (!pending) return false;

    clearTimeout(pending.timer);
    this.pending.delete(id);
    pending.reject(error);
    return true;
  }

  rejectAll(error: LionScraperError): void {
    for (const [id, pending] of this.pending) {
      clearTimeout(pending.timer);
      pending.reject(error);
      this.pending.delete(id);
    }
  }

  /** Reject all pending bridge calls with BRIDGE_DISCONNECTED, localized per request `lang`. */
  rejectAllDisconnected(reason: DisconnectRejectId): void {
    for (const [id, pending] of this.pending) {
      clearTimeout(pending.timer);
      pending.reject(
        createError(BridgeErrorCode.BRIDGE_DISCONNECTED, t(pending.lang, reason)),
      );
      this.pending.delete(id);
    }
  }

  has(id: string): boolean {
    return this.pending.has(id);
  }

  get size(): number {
    return this.pending.size;
  }

  /**
   * Clears pending timers and map entries without resolving or rejecting the returned Promises from {@link add}.
   * For production shutdown paths use {@link rejectAll} or {@link rejectAllDisconnected} instead.
   */
  clear(): void {
    for (const pending of this.pending.values()) {
      clearTimeout(pending.timer);
    }
    this.pending.clear();
  }
}

File: packages/python/src/lionscraper/bridge/protocol.py
from __future__ import annotations

import asyncio
import uuid
from typing import Any, Callable

from lionscraper.i18n.lang import SupportedLang, t
from lionscraper.types.bridge import BridgeMethod, BridgeProgressParams, BridgeRequest, BridgeResponse
from lionscraper.types.errors import BridgeErrorCode, LionScraperError, create_error
from lionscraper.utils.logger import logger

DEFAULT_TIMEOUT_MS = 60_000

BridgeProgressHandler = Callable[[BridgeProgressParams], None]


def create_bridge_request(method: BridgeMethod, params: dict[str, Any] | None = None) -> BridgeRequest:
    req: BridgeRequest = {
        "jsonrpc": "2.0",
        "id": str(uuid.uuid4()),
        "method": method,
    }
    if params is not None:
        req["params"] = params
    return req


def create_bridge_response(req_id: str, result: Any) -> BridgeResponse:
    return {"jsonrpc": "2.0", "id": req_id, "result": result}


def create_bridge_error_response(
    req_id: str,
    code: int,
    message: str,
    data: dict[str, Any] | None = None,
) -> BridgeResponse:
    err: dict[str, Any] = {"code": code, "message": message}
    if data:
        err["data"] = data
    return {"jsonrpc": "2.0", "id": req_id, "error": err}


DisconnectRejectId = str


class PendingRequest:
    __slots__ = ("id", "method", "fut", "timer_handle", "created_at", "lang", "on_progress")

    def __init__(
        self,
        id: str,
        method: str,
        fut: asyncio.Future[Any],
        timer_handle: asyncio.TimerHandle,
        lang: SupportedLang,
        on_progress: BridgeProgressHandler | None,
    ):
        self.id = id
        self.method = method
        self.fut = fut
        self.timer_handle = timer_handle
        self.lang = lang
        self.on_progress = on_progress


class PendingRequestManager:
    def __init__(self) -> None:
        self._pending: dict[str, PendingRequest] = {}

    def begin_request(
        self,
        req_id: str,
        method: BridgeMethod,
        timeout_ms: int = DEFAULT_TIMEOUT_MS,
        lang: SupportedLang = "en-US",
        on_progress: BridgeProgressHandler | None = None,
    ) -> asyncio.Future[Any]:
        loop = asyncio.get_running_loop()
        fut: asyncio.Future[Any] = loop.create_future()

        def on_timeout() -> None:
            self._pending.pop(req_id, None)
            if not fut.done():
                logger.warn(f"Bridge request timed out: {req_id} ({method})")
                fut.set_exception(
                    create_error(
                        BridgeErrorCode.BRIDGE_TIMEOUT,
                        t(lang, "bridge_timeout", {"ms": timeout_ms}),
                    )
                )

        handle = loop.call_later(timeout_ms / 1000.0, on_timeout)
        self._pending[req_id] = PendingRequest(req_id, method, fut, handle, lang, on_progress)
        return fut

    def dispatch_progress(self, request_id: str, params: BridgeProgressParams) -> bool:
        pending = self._pending.get(request_id)
        if not pending:
            return False
        if pending.on_progress:
            try:
                pending.on_progress(params)
            except Exception:
                pass
        return True

    def resolve(self, req_id: str, result: Any) -> bool:
        pending = self._pending.pop(req_id, None)
        if not pending:
            return False
        pending.timer_handle.cancel()
        if not pending.fut.done():
            pending.fut.set_result(result)
        return True

    def reject(self, req_id: str, error: LionScraperError) -> bool:
        pending = self._pending.pop(req_id, None)
        if not pending:
            return False
        pending.timer_handle.cancel()
        if not pending.fut.done():
            pending.fut.set_exception(error)
        return True

    def reject_all_disconnected(self, reason_key: str) -> None:
        for req_id, pending in list(self._pending.items()):
            pending.timer_handle.cancel()
            if not pending.fut.done():
                pending.fut.set_exception(
                    create_error(
                        BridgeErrorCode.BRIDGE_DISCONNECTED,
                        t(pending.lang, reason_key),
                    )
                )
            self._pending.pop(req_id, None)

    def has(self, req_id: str) -> bool:
        return req_id in self._pending

    @property
    def size(self) -> int:
        return len(self._pending)

    def clear(self) -> None:
        for p in self._pending.values():
            p.timer_handle.cancel()
        self._pending.clear()
